File: android/app/src/main/kotlin/run/nuri/getagrip/ui/components/FingerGlyph.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import run.nuri.getagrip.engine.FingerSet
import run.nuri.getagrip.engine.GripPosition
import run.nuri.getagrip.ui.theme.GetAGripTheme
import run.nuri.getagrip.ui.theme.LocalGripPalette

/// A grip drawn as four BARS — index to little, filled when that finger is on the edge and
/// hairline when it is not.
///
/// Bars, not dots, and that is the whole point. As squares with a full corner radius these
/// rendered as literal CIRCLES at open hand and drag — indistinguishable from the session
/// dots one row above them and the consistency strip below, so a card that meant three
/// different things looked like circles all the way down. A finger is taller than it is
/// wide; drawing it that way makes the collision impossible rather than merely unlikely, and
/// it matches the app icon, which is this same mark.
///
/// **ONE GLYPH, ONE RULE: full capsule, 22 : 38.** The radius is half the width, byte for
/// byte the same rule `PalmHand` and `HandMark` use, so the runner's hand, a History row and
/// a notification are recognisably one drawing. The closure-driven radius iOS once had is
/// GONE: half a radius is a subtler distinction than "same mark or not", and the position is
/// spoken in words everywhere the glyph appears.
///
/// TRANSLATION NOTE: iOS declares `dot`/`gap` as `@ScaledMetric` so the glyph grows with the
/// text beside it. Compose has no environment-scaled Dp; callers pass a plain size and the
/// system's font scale reaches the words next to it but not this. That is a real, small
/// difference from iOS — noted rather than faked, because scaling a drawing by
/// `LocalDensity.fontScale` would grow it in places iOS does not.
@Composable
fun FingerGlyph(
    fingers: FingerSet,
    modifier: Modifier = Modifier,
    position: GripPosition = GripPosition.halfCrimp,
    dot: Dp = 5.5.dp,
    gap: Dp = 3.dp,
    tint: Color = LocalGripPalette.current.graphite,
) {
    val palette = LocalGripPalette.current
    // Fingers are taller than they are wide — the ratio is what stops a bar ever reading as
    // a dot, at any size.
    val barHeight = dot * BAR_RATIO
    val radius = dot / 2

    Column(
        modifier = modifier.clearAndSetSemantics {},
        verticalArrangement = Arrangement.spacedBy(maxOf(1.5.dp, gap * 0.6f)),
        horizontalAlignment = Alignment.Start,
    ) {
        // BOTTOM-aligned, so the varying lengths hang from a common knuckle line the way
        // fingers do — top-aligned they splay downward and read as a chart.
        Row(
            horizontalArrangement = Arrangement.spacedBy(gap),
            verticalAlignment = Alignment.Bottom,
        ) {
            fingers.occupied.forEachIndexed { index, on ->
                val shape = RoundedCornerShape(radius)
                Box(
                    Modifier
                        .width(dot)
                        .height(barHeight * LENGTH_FACTOR[index])
                        .then(
                            if (on) {
                                Modifier.background(tint, shape)
                            } else {
                                // Stroked rather than a lighter fill: fill-vs-outline survives
                                // greyscale, Reduce Transparency and colourblindness; a tint
                                // step does not. FULL tertiary ink, no opacity — a 1 dp
                                // hairline is nearly all antialiased edge, and which fingers
                                // are OFF is half the grip's meaning.
                                Modifier.border(1.dp, palette.inkTertiary, shape)
                            },
                        ),
                )
            }
        }
        if (fingers.hasThumb) {
            // The thumb bar: horizontal, under the index side, because that is where a thumb
            // sits when a hand pinches. It keeps the shape grammar — BARS are digits; this
            // one just lies down.
            Box(
                Modifier
                    .width(dot * 2 + gap)
                    .height(dot * 0.62f)
                    .background(tint, RoundedCornerShape(radius)),
            )
        }
    }
}

/// The proportion that makes a bar a finger rather than a dot.
private const val BAR_RATIO = 1.75f

/// A HAND's proportions — middle longest, little shortest — matching `PalmHand` and
/// `HandMark`. Flat bars read as a barcode; these read as a hand at a glance, which is what
/// makes the glyph work at 6 dp in a History row.
private val LENGTH_FACTOR = listOf(0.86f, 1.0f, 0.94f, 0.80f)

@Preview(name = "FingerGlyph", showBackground = true)
@Composable
private fun FingerGlyphPreview() {
    GetAGripTheme {
        Row(
            Modifier.size(240.dp, 60.dp),
            horizontalArrangement = Arrangement.spacedBy(16.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            FingerGlyph(FingerSet.four, dot = 14.dp, gap = 6.dp)
            FingerGlyph(FingerSet.frontTwo, dot = 14.dp, gap = 6.dp)
            FingerGlyph(FingerSet.backTwo, dot = 14.dp, gap = 6.dp)
            FingerGlyph(
                FingerSet.frontTwo.union(FingerSet.thumb),
                position = GripPosition.pinch,
                dot = 14.dp,
                gap = 6.dp,
            )
        }
    }
}

File: Sources/UI/Components/FingerGlyph.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import SwiftUI

/// A grip drawn as four BARS — index to little, filled when that finger is on the edge
/// and hairline when it is not.
///
/// Bars, not dots: as full-radius squares these rendered as CIRCLES, indistinguishable
/// from the session dots above and the consistency strip below. A finger is taller than
/// wide; drawing it so makes the collision impossible, and matches the app icon.
///
/// The app's only iconography under the no-emoji rule, and **the same drawing as the
/// hand hanging off the Dynamic Island** — see `radius`.
struct FingerGlyph: View {
    let fingers: FingerSet
    var position: GripPosition = .halfCrimp
    /// `@ScaledMetric`, so the glyph grows with its text instead of shrinking to a speck.
    /// Callers pass the design size (`dot: 12`) and get it scaled.
    @ScaledMetric(relativeTo: .caption) var dot: CGFloat = 5.5
    @ScaledMetric(relativeTo: .caption) var gap: CGFloat = 3
    var tint: Color = Accent.graphite

    /// Taller than wide — `HandGeometry.barAspect` stops a bar ever reading as a dot.
    private var barHeight: CGFloat { dot * HandGeometry.barAspect }

    /// **A FULL CAPSULE — the island hand's rule exactly** (`IslandHand.radius`; Nuri,
    /// 2026-08-09: "identical radii so it's clear that they're the same fingers").
    ///
    /// It used to square off with `GripPosition.closure`, which separated the two front-2
    /// sets at a glance. That signal is spent on purpose: one mark drawn one way everywhere
    /// beats a hint only its author could read. The position is spoken in words wherever
    /// the glyph appears.
    private var radius: CGFloat { dot / 2 }

    var body: some View {
        VStack(alignment: .leading, spacing: max(1.5, gap * 0.6)) {
            // BOTTOM-aligned from a common knuckle line, like fingers; top-aligned they
            // read as a chart.
            HStack(alignment: .bottom, spacing: gap) {
                ForEach(Array(fingers.occupied.indices), id: \.self) { index in
                    pip(on: fingers.occupied[index], index: index)
                }
            }
            if fingers.hasThumb {
                // The thumb: horizontal, under the index side, where a pinching thumb sits
                // (Nuri: "a little rectangle under the 4"). BARS are digits; this one lies down.
                RoundedRectangle(cornerRadius: radius, style: .continuous)
                    .fill(tint)
                    .frame(width: dot * 2 + gap, height: dot * 0.62)
            }
        }
        // The parent speaks the whole grip as one sentence (`GripSpec.spoken`).
        .accessibilityHidden(true)
    }

    @ViewBuilder
    private func pip(on: Bool, index: Int) -> some View {
        let shape = RoundedRectangle(cornerRadius: radius, style: .continuous)
        Group {
            if on {
                shape.fill(tint)
            } else {
                // Stroked, not a lighter fill: fill-vs-outline survives greyscale, Reduce
                // Transparency and colourblindness. FULL tertiary, measured: 0.45 gave 2.3:1
                // and 0.7 only 2.6:1 on the dark well, since a 1 pt hairline is nearly all
                // antialiased edge. Which fingers are OFF is half the grip; it must clear 3:1.
                shape.strokeBorder(Ink.tertiary, lineWidth: 1)
            }
        }
        .frame(width: dot, height: barHeight * HandGeometry.lengthFactor[index])
    }
}

#Preview {
    VStack(alignment: .leading, spacing: 18) {
        ForEach(GripPosition.known, id: \.rawValue) { position in
            HStack(spacing: 16) {
                FingerGlyph(fingers: .four, position: position, dot: 14, gap: 6)
                FingerGlyph(fingers: .frontThree, position: position, dot: 14, gap: 6)
                FingerGlyph(fingers: .frontTwo, position: position, dot: 14, gap: 6)
                FingerGlyph(fingers: .backTwo, position: position, dot: 14, gap: 6)
                Text(position.name)
                    .font(.system(.footnote, weight: .medium))
                    .foregroundStyle(Ink.secondary)
            }
        }
    }
    .padding()
    .background { AppBackground() }
}
